Documentation/rfc/v3api.md

```diff
@@ -0,0 +1,459 @@
+  optional int64 index = 4;
+  optional uint64 raft_term = 5;
+}
+
+message RangeRequest {
+  optional bytes key = 1;
+  // if the end_key is not given, it is a get.
+  // if the end_key is given, it gets the keys in range [key, end_key).
+  optional bytes end_key = 2;
+  // limit the number of keys returned.
+  optional int64 limit = 3;
+}
+
+message RangeResponse {
+  optional ResponseHeader header = 1;
+  repeated KeyValue kvs = 2;
+}
+
+message PutRequest {
+  optional bytes key = 1;
+  optional bytes value = 2;
+}
+
+message PutResponse {
+  optional ResponseHeader header = 1;
+}
+
+message DeleteRequest {
+  optional bytes key = 1;
+}
+
+message DeleteResponse {
+  optional ResponseHeader header = 1;
+}
+
+message RequestUnion {
+  oneof value {
+    RangeRequest request_range = 1;
+    PutRequest request_put = 2;
+    DeleteRequest request_delete = 3;
+  }
+}
+
+message ResponseUnion {
+  oneof value {
+    RangeResponse request_range = 1;
+    PutResponse request_put = 2;
+    DeleteResponse request_delete = 3;
+  }
+}
+
+message Compare {
+  enum CompareType {
+    equal = 0;
+    greater = 1;
+    less = 2;
+  }
+  optional CompareType type = 1;
+  // key path
+  optional bytes key = 2;
+  one of {
+      // version of the given key
+      int64 version = 1;
+    // create index of the given key
+    int64 create_index = 2;
+    // last modified index of the given key
+    int64 mod_index = 3;
+    // value of the given key
+    bytes value = 4;
+  }
+}
+
+// First all the compare requests are processed.
+// If all the compare succeed, all the success
+// requests will be processed.
+// Or all the failure requests will be processed and
+// all the errors in the comparison will be returned.
+
+// From google paxosdb paper:
+// Our implementation hinges around a powerful primitive which we call MultiOp. All other database
+// operations except for iteration are implemented as a single call to MultiOp. A MultiOp is applied atomically
+// and consists of three components:
+// 1. A list of tests called guard. Each test in guard checks a single entry in the database. It may check
+// for the absence or presence of a value, or compare with a given value. Two different tests in the guard
+// may apply to the same or different entries in the database. All tests in the guard are applied and
+// MultiOp returns the results. If all tests are true, MultiOp executes t op (see item 2 below), otherwise
+// it executes f op (see item 3 below).
+// 2. A list of database operations called t op. Each operation in the list is either an insert, delete, or
+// lookup operation, and applies to a single database entry. Two different operations in the list may apply
+// to the same or different entries in the database. These operations are executed
+// if guard evaluates to
+// true.
+// 3. A list of database operations called f op. Like t op, but executed if guard evaluates to false.
+message TnxRequest {
```

Comment:
Just going to throw this out there: is it maybe reasonable to have TxnRequest  be  `{ repeated MultiOp ops =1; }` -- this gets you the equivalent of a case statement/if-else chain etc.